protos/idl/service.proto

```diff
@@ -98,3 +116,74 @@ message Tag {
 message Metadata {
     map<string, string> key_map = 1; // key map is a dictionary of key/val strings that represent metadata
 }
+
+// Filter expression that is composed of a combination of single filters
+message FilterExpression {
+    repeated SinglePropertyFilter filters = 1;
+}
+
+// A single property to filter on.
+message SinglePropertyFilter {
+    oneof property_filter {
+        TagPropertyFilter tag_filter = 1;
+        PartitionPropertyFilter partition_filter = 2;
+        ArtifactPropertyFilter artifact_filter = 3;
+        DatasetPropertyFilter dataset_filter = 4;
+    }
+
+    // as use-cases come up we can add more operators, ex: gte, like, not eq etc.
+    enum ComparisonOperator {
+        EQUALS = 0;
+    }
+
+    ComparisonOperator operator = 10; // field 10 in case we add more entities to query
+    // Next field number: 11
+}
+
+// Artifact properties we can filter by
+message ArtifactPropertyFilter {
+    oneof property {
```

Comment:
Just curious why you went with oneofs? Do you anticipate adding more fields here in the future?

Reply:
Yea, it is for future proofing. We anticipate that there will be more properties we want to filter on later. I can add a comment for that.